driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,731 @@
+        String invalidConnectionAccessToken = connectionLastAccessToken;
+        String cachedRefreshToken = cacheEntry.getRefreshToken();
+        IdpInfo cachedIdpInfo = cacheEntry.getIdpInfo();
+
+        if (cachedAccessToken != null) {
+            boolean cachedTokenIsInvalid = cachedAccessToken.equals(invalidConnectionAccessToken);
+            if (cachedTokenIsInvalid) {
+                mongoCredentialWithCache.setOidcCacheEntry(cacheEntry.clearAccessToken());
+                cachedAccessToken = null;
+            }
+        }
+        OidcRefreshCallback refreshCallback = getRefreshCallback();
+        if (cachedAccessToken != null) {
+            fallbackState = FallbackState.PHASE_1_CACHED_TOKEN;
+            return prepareTokenAsJwt(cachedAccessToken);
+        } else if (refreshCallback != null && cachedRefreshToken != null) {
+            assertNotNull(cachedIdpInfo);
+            // Invoke Refresh Callback using cached Refresh Token
+            validateAllowedHosts(getMongoCredential());
+            fallbackState = FallbackState.PHASE_2_REFRESH_CALLBACK_TOKEN;
+            IdpResponse result = refreshCallback.onRefresh(new OidcRefreshContextImpl(
+                    cachedIdpInfo, cachedRefreshToken, CALLBACK_TIMEOUT));
+            return populateCacheWithCallbackResultAndPrepareJwt(cachedIdpInfo, result);
+        } else {
+            // cache is empty
+
+            /*
+            A check for present idp info short-circuits phase-3a.
+
+            If a challenge is present, it can only be a response to a
+            "principal-request", so the challenge must be the resulting
+            idp info. Such a request is made during speculative auth,
+            though the source is unimportant, as long as we detect and
+            use it here.
+
+            Checking that the fallback state is not phase-3a ensures that
+            this does not loop infinitely in the case of a bug.
+             */
+            boolean idpInfoNotPresent = challenge.length == 0;
+            if (fallbackState != FallbackState.PHASE_3A_PRINCIPAL && idpInfoNotPresent) {
+                fallbackState = FallbackState.PHASE_3A_PRINCIPAL;
+                return prepareUsername(mongoCredentialWithCache.getCredential().getUserName());
+            } else {
+                IdpInfo idpInfo = toIdpInfo(challenge);
+                validateAllowedHosts(getMongoCredential());
+                IdpResponse result = requestCallback.onRequest(new OidcRequestContextImpl(idpInfo, CALLBACK_TIMEOUT));
+                fallbackState = FallbackState.PHASE_3B_REQUEST_CALLBACK_TOKEN;
+                return populateCacheWithCallbackResultAndPrepareJwt(idpInfo, result);
+            }
+        }
+    }
+
+    private boolean isAutomaticAuthentication() {
+        return getRequestCallback() == null;
+    }
+
+    private boolean clientIsComplete() {
+        return fallbackState != FallbackState.PHASE_3A_PRINCIPAL;
+    }
```

Comment:
Check that this method is not just a wrapper around a constant (see https://github.com/mongodb/mongo-java-driver/pull/1259#discussion_r1490268147).

Reply:
It is no longer returning `true` as in the link, but looks as it does here.